[WIP] Playing around with grid persistence.

csarofeen opened the pull request:
Made some changes to the grid persistent branch trying to get perf starting with an arbitrary size I would hope would be straight forward to get perf on (not really certain it is). I hardcoded this size under the benchmark: `NvFuserScheduler_ResNext_BatchNorm_nhwc_fp16`.

zasdfgbnm committed:
Refactor TransormPropagator to allow specifying a position and propagating to part of the DAG (#1775)

`MaxInfoPropagator` is renamed to `MaxInfoSpanningTree`, it now only does path-finding, and the propagation is in a separate class `MaxInfoSpanningTree::Propagator`. Same for `MaxRootDomainInfoPropagator`.

`MaxInfoSpanningTree` and `MaxRootDomainInfoSpanningTree`  now allow specifying a selector, which controls which subgraph should be included in path-finding.

`MaxRootDomainInfoSpanningTree` also gets a few new constructors for convenience to use.

`TransormPropagator` is now a subclass of `MaxInfoSpanningTree::Propagator`, so the way to use it has changed.

Now `MaxInfoSpanningTree` and `MaxRootDomainInfoSpanningTree` will store the path after generation so that the same path can be traversed multiple times. This will be useful to support use cases like new `computeAt`. Pseudo-code:
```C++
void TensorView::computeAt(TensorView tv, int pos) {
  auto ComputeAtSubgraphSelector selector(this, tv);
  MaxRootDomainInfoSpanningTree path(tv, pos, &selector);
  TransformPropagator propagator(tv, pos);
  path.traverse(&propagator);
  ComputeAtPosPropagator ca_propagator(tv, pos);
  path.traverse(&ca_propagator);
}
```

csarofeen committed:
Revert scheduling changes. Cleanup only.

csarofeen committed:
Start drafting grid persistent kernels.

csarofeen commented:
I'm getting 142GB/s on this size right now, which is underwhelming, but is better than where it started at.

csarofeen commented on benchmarks/cpp/nvfuser/batch_norm_channels_last.cpp:
```diff
@@ -298,16 +298,16 @@ NVFUSER_BENCHMARK_DEFINE(
+    // ->Args({128, 512, 14})
+    // ->Args({128, 1024, 14})
+    // ->Args({128, 1024, 7})
+    ->Args({8, 4096, 64})
```
TODO: Remove before considering to merge.

csarofeen commented on torch/csrc/jit/codegen/cuda/codegen.cpp:
```diff
@@ -1963,6 +1963,11 @@ class CudaKernelGenerator : private OptOutConstDispatch {
     const auto reduction_name =
         genFusedReductionName(alloc_fused_reduction->out()->view());
 
+    if (!alloced_reduction_classes.emplace(alloc_fused_reduction->out()->view())
```
With welford, in an unswitch section there was multiple definitions being created because the grid reduce was being used in both the then/else section of the code. We should go fix the actual insertion of the allocations, instead of WAR-ing the codegen.

csarofeen commented on torch/csrc/jit/codegen/cuda/fusion_segmenter.h:
```diff
@@ -411,7 +411,8 @@ class CombineReductions;
 struct TORCH_CUDA_CU_API SegmentCandidateFinderOptions {
-  bool run_translate_welford = true;
+  // Temporarily disable by default for now. May want to revisit.
+  bool run_translate_welford = false;
```
We want to use welford for grid persistence, forcing translation off everywhere for now.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -171,15 +171,6 @@ ReductionParams innerPersistentHeuristic(
   const int64_t max_multi_reduction_factor = scheduler_utils::safeDiv(
       scheduler_utils::register_file_size, max_persistent_buffer_size);
 
```
Removing inner persistent changes for now.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -483,45 +432,25 @@ ReductionParams innerPersistentHeuristic(
   if (godim > 1) {
-    if (rparams.cross_grid_inner_reduction) {
-      rparams.grid_dim_iter_dom = ParallelType::BIDy;
+    rparams.grid_dim_iter_dom = ParallelType::BIDx;
```
Should have kept this change instead of reverting. Can we isolate this into a stand alone PR as it seems to be a fix.

naoyam replied:
As the grid persistence is revered, doesn't this change make any difference? Am I missing something?

csarofeen replied:
I was trying to just remove the changes unrelated to outer persistent reductions since that's what we really need quickly.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -593,7 +524,7 @@ ReductionParams outerPersistentHeuristic(
       // Reduce unrolling if we have many inputs, start reduction at 4 inputs
       scheduler_utils::lastPow2(
-          scheduler_utils::safeDiv((int64_t)n_tensor_inputs, 2)));
+          scheduler_utils::safeDiv((int64_t)n_tensor_inputs, 4)));
```
`<<2` is the equivalent of dividing by `4` not `2`.

naoyam replied:
😳

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -711,22 +645,25 @@ ReductionParams outerPersistentHeuristic(
-         num_warp_per_partition) {
-    bdimy++;
-  }
+  //   // Increase bdimy until we hit the SM partition granularity
```
I removed this, only because I didn't understand what you're trying to do here. On second thought, I assume you're just trying to avoid going across warp boundaries that would allocate more registers, but uncertain why.

csarofeen replied:
I agree we probably want to consider these boundaries, but we're already limiting the register file quite significantly (more so than this could inflate the registers by).

naoyam replied:
This was just copied over from your original PR. I thought that it was meant to increase the number of threads within the limit of the `32 * #partitions` granularity as doing so would have no actual impact on resource usage.

csarofeen replied:
That's probably right, we can revisit after we get a bit more out of the current available perf.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -793,24 +734,30 @@ ReductionParams outerPersistentHeuristic(
+  std::cout << "7: " << bdimx << ", " << bdimy << std::endl;
+  // Disable for now, it can go from taking half a warp, to reducing it to a
+  // quarter warp. On cases where we're dealing with fp16 values and we're not
+  // unrolled, this would mean loads are only 8xfp16 words strided across tidy.
```
This ^^

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -819,28 +766,33 @@ ReductionParams outerPersistentHeuristic(
+  // Set gdimy.
+  if (!block_persist) {
+    gdimy = sm_required_per_norm_set;
+  }
```
Need to set gdimy before trying to figure out the size of gdimx

naoyam replied:
Yeah, thanks for the fix.

naoyam commented:
Thanks! I saw performance regressions when disabled welfrod translations. It seemed it was actually better not to translate for outer-reduction cases, but for inner-reduction batch norms, I saw up to 20% regressions. I'll report it back after doublechecking what I saw.

shmsong committed:
Extend mma dimension and layout checking to support strided batched matmul and tensor contractions (#1761)

Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

zasdfgbnm committed:
Fix FusionMaxRootDomainInfoSpanningTreePrintTwice_CUDA (#1781)

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:
```diff
@@ -1243,30 +1237,39 @@ class PersistentKernelScheduler : public SchedulerEntry {
         ceilDiv(persistent_buffer_size, scheduler_utils::register_file_size);
 
+    // If the persistence requires over half the device don't do grid
+    // persistence as we can't overlap the grid comms.
```
What is overlapped with the grid comms?

csarofeen replied:
Just different groups of blocks running at the same time. I just don't think we want all SM's participating in the same grid persistent allreduce

IvanYashchuk committed:
Fix div(Val, TensorView) (#1778)

* Fix div(scalar, tensor)

* lintrunner: clang-format

naoyam reviewed (approved):
Thanks for the PR.

naoyam commented:
> I'm getting 142GB/s on this size right now, which is underwhelming, but is better than where it started at.

Do you know what is the bandwidth with the non-persistent scheduler?

zasdfgbnm committed:
Adding sibling path for MaxInfoSpanningTree (#1776)

The sibling path is required to generate consistent replay for some cases where `MaxInfoSpanningTree` is used with a selector. For example, when the producer of a Welford is excluded from the propagation section. See test `FusionTransformPropagateSelectorSibling_CUDA` for a detailed example. Besides, since we know that siblings should be transformed exactly the same, the sibling path is a perfect next hop for preserving information.

If you want a spanning tree without a sibling path, you can override `allowSibling` as `return false` in your selector;

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp:
```diff
@@ -711,22 +645,25 @@ ReductionParams outerPersistentHeuristic(
 
+  std::cout << "4: " << bdimx << ", " << bdimy << std::endl;
   // Move unrolling factor into vectorization upto vectorization limit.
   if (vectorize_factor > 1 && iter_unroll_factor > 1) {
```
This block needs to be protected because it's setting `iter_unroll_factor` but not checking `max_multi_reduction_factor`. We need to change this block to:
```
  if (vectorize_factor > 1 && iter_unroll_factor > 1) {
    vectorize = true;
    iter_unroll_factor = std::min(
        scheduler_utils::lastPow2(iter_unroll_factor),
        (int64_t)vectorize_factor);
    iter_unroll_factor = std::min(
        iter_unroll_factor,
        scheduler_utils::safeDiv(max_multi_reduction_factor, bdimx));
  }
  ```

csarofeen commented:
For some reason I seem to not have a response block to your inline comments:

> As the grid persistence is revered, doesn't this change make any difference? Am I missing something?

Just ignore any changes I made to inner persistent for now. I was just trying to isolate out cleanup from heuristic changes. Just focus on the outer persistent stuff, and you can push fixes to inner persistent in another PR.

> This was just copied over from your original PR. I thought that it was meant to increase the number of threads within the limit of the 32 * #partitions granularity as doing so would have no actual impact on resource usage.

Yeah I think you're right, let's revisit this later to see if it's actually valuable.



>Do you know what is the bandwidth with the non-persistent  scheduler?

Let me pull it.

csarofeen committed:
Save.

csarofeen commented:
With grid persistence:
kernel1 run in 4.37616 ms, achieved: 122.7 GB/s

Without:
kernel1 run in 1.08733 ms, achieved: 246.952 GB/s
kernel2 run in 0.892608 ms, achieved: 601.5 GB/s

csarofeen commented:
there's definitely some register usage issues. I think we need to check on `nvfuser_zero` insertion, as manually adding more of those and zero update was reducing register pressure.

shmsong committed:
Disable register reuse across serial broadcast ops (#1787)

Disable memory aliasing for inner sharing across serial broadcast.

Sergei-Lebedev committed:
Fix isIntegralType error msg (#1789)

zasdfgbnm committed:
Transform propagator skip replay when possible (#1782)

This comment in the code describes what this PR is doing:

```C++
  // Note: [Using multiple TransformPropagators]
  // There are cases that we use multiple TransformPropagators along different
  // spanning trees with different references in the same fusion. Some of these
  // spanning trees could overlap. In cases when there are overlapping nodes,
  // TransformPropagator needs to respect the replay of others, because the
  // current TransformPropagator might not contain the most amount of
  // information on how to do the correct transformation. The logic below tells
  // TransformPropagator to skip the replay when not necessary.
```

jjsjann123 committed:
Output allocate patch (#1790)

Caching strides along with sizes. This is to support current expand, which introduces non-contiguous output tensor

zasdfgbnm committed:
Add SpanningTreePrinter (#1786)

zasdfgbnm committed:
New compute at interface (#1743)

Rewrite of the compute at pass to rely on the new propagation mechanisms.

zasdfgbnm committed:
Fix TransformReplay::getMatchedLeafPosWithoutReplay* (#1791)

zasdfgbnm committed:
Some further cleanup for the new computeAt interface (#1793)

Revert MaxProducerPosUpdater to old algo.

zasdfgbnm committed:
Use TransformPropagatorWithCheck in many tests (#1795)

zasdfgbnm committed:
validateDomain in TransformPropagator (#1796)

zasdfgbnm committed:
InlinePropagator please don't replay (#1797)

This PR makes `InlinePropagator` just set compute-at positions. It will not replay any tensor. If you want to replay, please use `TransformPropagator` and friends to do so.

Currently, `InlinePropagator` is already asserting no replay for standard and best effort compute at. So this PR is mostly about making most inlined compute at works as well.

This PR also does a lot of cleanups to remove the word "replay" from comments and variable and function names from `InlinePropagator`.

I also cleaned up `recordReplayedPos` and `retrieveReplayedPos`, now the logic is much easier to understand.

zasdfgbnm committed:
Coding style cleanups (#1798)

Per offline discussion with @csarofeen, this PR does many renaming for better coding style: For all propagation-related things, I am now using the names `P2C` and `C2P` instead of `CasP` and `PasC`. Because "A as B" somewhat implies we want to replay A the same as B, but "B to A" sounds more general and is a better word for this case. Also, I modified the order of function arguments to match the order in its name. For example `PasC` should have `(producer, consumer)` or `(to, from)`, but not `(consumer, producer)` or `(from, to)`, and `C2P` should have `(consumer, producer)` or `(from, to)`, but not `(producer, consumer)` or `(to, from)`.

kevinstephano committed:
Add parsing support for `_to_copy` to handle AMP casts. (#1756)

1. Add support for _to_copy() to support AMP casts.
2. refactored cast, accept none for dtype
3. python tests

Co-authored-by: jjsjann123 <jiej@nvidia.com>

shmsong committed:
MMA Rfactor support for cross-warp and cross-CTA split on K dimension (#1554)

shmsong committed:
Indexing refactor stage 2 : Remove reference tensor in predicate indexing logic (#1784)

Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

zasdfgbnm committed:
More cleanup on InlinePropagator (#1800)

I just realized that `InlinePropagator` can be further simplified because it no longer replays.

Since `InlinePropagator` is no longer doing replay, it is more like a "for each" problem rather than a propagation problem:

For each tensor `tv`, if we already know what is the max position of `tv` that is mapped to the reference tensor's selected outer dimensions(stored in `mapped_reference_pos_` in the code), setting the CA position is a very local operation, and is as simple as checking `tv` itself and all its consumers to determine the inline position.

`InlinePropagator` is not completely a "for each" problem only because the computation of `mapped_reference_pos_` is a propagation problem.

This cleanup reorganizes the code of `InlinePropagator` so it is clear that `InlinePropagator` is nothing but a two-step process:
Step 1: Do a propagation to find the `mapped_reference_pos_` for all tensors.
Step 2: For each tensor, check itself and its consumers to determine the CA position.

Conceptually, I would like to split step 1 with step 2. Because this split makes these concepts decoupled. Especially, this PR makes `mapped_reference_pos_` only contain info about the reference tensor, and is independent of the CA position (Currently, this is not true for best effort and most inlined computeAt without this PR). Now, in my view, `InlinePropagator` is conceptually very simple and easy to understand.

In terms of implementation, step 1 and step 2 can be interleaved, because when we don't need to know the `mapped_reference_pos_` for `tv`'s consumer in order to compute the CA position of `tv`. So a one-pass traverse could do both step 1 and step 2 altogether.

shmsong committed:
Temporarily disable test requring large shared memory. (#1802)

naoyam committed:
Grouping grid allreduces across iterations (#1755)

* Extend the grouped grid reduction kernel

The kernel itself should work with an arbitrary number of inputs, but
the underlying data structure, Tuple, still explicitly needs to be
specialized for the number of values, which is currently limited to 8.

zasdfgbnm committed:
Check siblings in getMaxPosAll (#1805)

shmsong committed:
remove dead indexing code (#1806)

jjsjann123 committed:
Broadcast in dim with expand (#1794)

Fixes #1788

Added expand in broadcast_in_dim to support expanding to concrete size. Note that we are not supporting dynamic shape for concrete size at this moment.

jjsjann123 committed:
spam nvrtc options (#1783)

TORCH_WARN on nvrtc debug option impacting performance.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into grid_persistent_normalization_cs
